File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/statistikk/sykefraværshistorikk/SykefraværForEttKvartal.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk;

import com.fasterxml.jackson.annotation.JsonIgnore;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.ÅrstallOgKvartal;

import java.math.BigDecimal;
import java.util.Comparator;
import java.util.Objects;

public class SykefraværForEttKvartal extends MaskerbartSykefravær implements
    Comparable<SykefraværForEttKvartal> {

  @JsonIgnore
  private final ÅrstallOgKvartal årstallOgKvartal;

  @JsonIgnore
  private final int antallPersoner; // TODO: trenger vi det?

  public SykefraværForEttKvartal(
      ÅrstallOgKvartal årstallOgKvartal,
      BigDecimal tapteDagsverk,
      BigDecimal muligeDagsverk,
      int antallPersoner
  ) {
    super(
        tapteDagsverk,
        muligeDagsverk,
        antallPersoner,
        årstallOgKvartal != null
    );
    this.årstallOgKvartal = årstallOgKvartal;
    this.antallPersoner = antallPersoner;
  }

  public int getKvartal() {
    return årstallOgKvartal != null ? årstallOgKvartal.getKvartal() : 0;
  }

  public int getÅrstall() {
    return årstallOgKvartal != null ? årstallOgKvartal.getÅrstall() : 0;
  }

  public int getAntallPersoner() {
    return antallPersoner;
  }

  @Override
  public int compareTo(SykefraværForEttKvartal sykefraværForEttKvartal) {
    return Comparator
        .comparing(SykefraværForEttKvartal::getÅrstallOgKvartal)
        .compare(this, sykefraværForEttKvartal);
  }

  public ÅrstallOgKvartal getÅrstallOgKvartal() {
    return årstallOgKvartal;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof SykefraværForEttKvartal)) {
      return false;
    }
    if (!super.equals(o)) {
      return false;
    }
    SykefraværForEttKvartal that = (SykefraværForEttKvartal) o;

    return (årstallOgKvartal.equals(that.årstallOgKvartal)
        && getProsent() == null ?
        that.getProsent() == null : getProsent().equals(that.getProsent())
        && getTapteDagsverk() == null ?
        that.getTapteDagsverk() == null : getTapteDagsverk().equals(that.getTapteDagsverk())
        && getMuligeDagsverk() == null ?
        that.getMuligeDagsverk() == null : getMuligeDagsverk().equals(that.getMuligeDagsverk())
    );
  }

  @Override
  public int hashCode() {
    return Objects.hash(super.hashCode(), årstallOgKvartal);
  }
}

File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/applikasjon/domenemodeller/SykefraværForEttKvartal.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller

import com.fasterxml.jackson.annotation.JsonIgnore
import java.math.BigDecimal
import java.util.*
import java.util.function.Function

open class SykefraværForEttKvartal(
    @field:JsonIgnore val ÅrstallOgKvartal: ÅrstallOgKvartal?,
    tapteDagsverk: BigDecimal?,
    muligeDagsverk: BigDecimal?,
    // TODO: trenger vi det?
    @field:JsonIgnore val antallPersoner: Int
) : MaskerbartSykefravær(
    tapteDagsverk,
    muligeDagsverk,
    antallPersoner,
    ÅrstallOgKvartal != null && tapteDagsverk != null && muligeDagsverk != null
), Comparable<SykefraværForEttKvartal> {

    val kvartal: Int
        get() = ÅrstallOgKvartal?.kvartal ?: 0
    val Årstall: Int
        get() = ÅrstallOgKvartal?.årstall ?: 0

    override fun compareTo(other: SykefraværForEttKvartal): Int {
        return Comparator.comparing(Function { obj: SykefraværForEttKvartal -> obj.ÅrstallOgKvartal })
            .compare(this, other)
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o !is SykefraværForEttKvartal) {
            return false
        }
        if (!super.equals(o)) {
            return false
        }
        val that = o
        val erÅrstallOgKvartalLike = ÅrstallOgKvartal == that.ÅrstallOgKvartal
        val erProsentLike =
            if (this.getProsent() == null) that.getProsent() == null else this.getProsent().equals(that.getProsent())
        val erTapteDagsverkLike =
            if (this.getTapteDagsverk() == null) that.getTapteDagsverk() == null else this.getTapteDagsverk()
                .equals(that.getTapteDagsverk())
        val erMuligeDagsverkLike =
            if (this.getMuligeDagsverk() == null) that.getMuligeDagsverk() == null else this.getMuligeDagsverk()
                .equals(that.getMuligeDagsverk())
        return erÅrstallOgKvartalLike && erProsentLike && erTapteDagsverkLike && erMuligeDagsverkLike
    }

    override fun hashCode(): Int {
        return Objects.hash(super.hashCode(), ÅrstallOgKvartal)
    }
}
